src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt

```diff
@@ -0,0 +1,164 @@
+        is ParameterType.Valid -> activeType.typeReference
+        else -> null
+    }
+    fun parsePat(): RsPat? = factory.tryCreatePat(patText)
+
+    val pat: RsPat
+        get() = parsePat() ?: factory.createPat("_")
+}
+
+/**
+ * This class holds information about function's properties (name, return type, parameters, etc.).
+ * It is designed to be changed (mutably) in the Change Signature dialog.
+ *
+ * After the dialog finishes, the refactoring will compare the state of the original function with the modified config
+ * and perform the necessary adjustments.
+ */
+class RsChangeFunctionSignatureConfig private constructor(
+    function: RsFunction,
+    var name: String,
+    val originalParameters: List<Parameter>,
+    var returnTypeDisplay: RsTypeReference?,
+    var visibility: RsVis? = null,
+    var isAsync: Boolean = false,
+    var isUnsafe: Boolean = false
+) : RsFunctionSignatureConfig(function) {
+    override fun typeParameters(): List<RsTypeParameter> = function.typeParameters
+
+    val returnTypeReference: RsTypeReference
+        get() = returnTypeDisplay ?: RsPsiFactory(function.project).createType("()")
+
+    val allowsVisibilityChange: Boolean
+        get() = !(function.owner is RsAbstractableOwner.Trait || function.owner.isTraitImpl)
+
+    val parameters: MutableList<Parameter> = originalParameters.toMutableList()
+
+    private val originalName: String = function.name.orEmpty()
+
+    val returnType: Ty
+        get() = returnTypeDisplay?.type ?: TyUnit
+
+    private val parametersText: String
+        get() {
+            val self = function.selfParameter?.text.orEmpty()
+            val prefix = if (self.isNotEmpty()) ", " else ""
+            val parameters = parameters.joinToString(", ", prefix = prefix) { "${it.pat.text}: ${it.typeReference.text}" }
+            return "${self}$parameters"
+        }
+
+    fun signature(): String = buildString {
+        visibility?.let { append("${it.text} ") }
+
+        if (isAsync) {
+            append("async ")
+        }
+        if (isUnsafe) {
+            append("unsafe ")
+        }
+        append("fn $name$typeParametersText($parametersText)")
+        if (returnType !is TyUnit) {
+            append(" -> ${returnTypeReference.text}")
+        }
+        append(whereClausesText)
+    }
+
+    fun createChangeInfo(): ChangeInfo = RsSignatureChangeInfo(this)
+    fun nameChanged(): Boolean = name != originalName
+    fun parameterSetOrOrderChanged(): Boolean = parameters.map { it.index } != originalParameters.indices.toList()
+
+    companion object {
+        fun create(function: RsFunction): RsChangeFunctionSignatureConfig {
+            val parameters = function.valueParameters.mapIndexed { index, parameter ->
+                Parameter(RsPsiFactory(function.project), parameter.pat?.text
+                    ?: "_", ParameterType.fromTypeReference(parameter.typeReference), index)
```

Comment:
```suggestion
                val factory = RsPsiFactory(function.project)
                val patText = parameter.pat?.text ?: "_"
                val type = ParameterType.fromTypeReference(parameter.typeReference)
                Parameter(factory, patText, type, index)
```

Or at least, can we please have line break after comma, not in the middle of argument